File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/config/CookieRequestCacheTest.java
package uk.gov.justice.digital.hmpps.oauth2server.config;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.util.Base64Utils;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.*;

@ExtendWith(SpringExtension.class)
class CookieRequestCacheTest {
    @Mock
    private SavedRequestCookieHelper helper;
    @Mock
    private HttpServletRequest request;
    @Mock
    private HttpServletResponse response;
    @Captor
    private ArgumentCaptor<String> captor;

    private CookieRequestCache cache;

    @BeforeEach
    void setUp() {
        cache = new CookieRequestCache(helper);
    }

    @Test
    void saveRequest_secureRequest() {
        when(request.getRequestURL()).thenReturn(new StringBuffer("http://some.com:443/where"));
        when(request.getQueryString()).thenReturn("param=value");
        when(request.isSecure()).thenReturn(Boolean.TRUE);

        cache.saveRequest(request, response);

        verify(helper).addCookieToResponse(same(request), same(response), captor.capture());

        final var url = new String(Base64Utils.decodeFromString(captor.getValue()));
        assertThat(url).isEqualTo("https://some.com/where?param=value");
    }

    @Test
    void saveRequest_insecureRequest() {
        when(request.getRequestURL()).thenReturn(new StringBuffer("https://some.com/where"));
        when(request.getQueryString()).thenReturn("param=value");
        when(request.isSecure()).thenReturn(Boolean.FALSE);

        cache.saveRequest(request, response);

        verify(helper).addCookieToResponse(same(request), same(response), captor.capture());

        final var url = new String(Base64Utils.decodeFromString(captor.getValue()));
        assertThat(url).isEqualTo("http://some.com/where?param=value");
    }

    @Test
    void saveRequest_differentPort() {
        when(request.getRequestURL()).thenReturn(new StringBuffer("https://some.com:12345/where"));
        when(request.getQueryString()).thenReturn("param=value");
        when(request.isSecure()).thenReturn(Boolean.TRUE);

        cache.saveRequest(request, response);

        verify(helper).addCookieToResponse(same(request), same(response), captor.capture());

        final var url = new String(Base64Utils.decodeFromString(captor.getValue()));
        assertThat(url).isEqualTo("https://some.com:12345/where?param=value");
    }

    @Test
    void getRequest() {
        when(helper.readValueFromCookie(request)).thenReturn(
                Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".getBytes())));

        final var savedRequest = cache.getRequest(request, response);

        assertThat(savedRequest.getRedirectUrl()).isEqualTo("https://some.com/where?param=value");
    }

    @Test
    void getMatchingRequest_matches() {
        when(request.getRequestURL()).thenReturn(new StringBuffer("http://some.com:443/where"));
        when(request.getQueryString()).thenReturn("param=value");
        when(request.isSecure()).thenReturn(Boolean.TRUE);

        when(helper.readValueFromCookie(request)).thenReturn(
                Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".getBytes())));

        final var savedRequest = cache.getMatchingRequest(request, response);

        assertThat(savedRequest).isSameAs(request);

        verify(helper).removeCookie(request, response);
    }

    @Test
    void getMatchingRequest_nomatch() {
        when(request.getRequestURL()).thenReturn(new StringBuffer("http://some.com:443/where"));
        when(request.getQueryString()).thenReturn("param=othervalue");
        when(request.isSecure()).thenReturn(Boolean.TRUE);

        when(helper.readValueFromCookie(request)).thenReturn(
                Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".getBytes())));

        final var savedRequest = cache.getMatchingRequest(request, response);

        assertThat(savedRequest).isNull();

        verify(helper, never()).removeCookie(request, response);
    }

    @Test
    void getMatchingRequest_noSavedRequest() {
        when(helper.readValueFromCookie(request)).thenReturn(Optional.empty());

        final var savedRequest = cache.getMatchingRequest(request, response);

        assertThat(savedRequest).isNull();
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/config/CookieRequestCacheTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.config

import com.nhaarman.mockitokotlin2.check
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.same
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.springframework.util.Base64Utils
import java.util.Optional
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class CookieRequestCacheTest {
  private val helper: SavedRequestCookieHelper = mock()
  private val request: HttpServletRequest = mock()
  private val response: HttpServletResponse = mock()
  private val cache = CookieRequestCache(helper)

  @Test
  fun saveRequest_secureRequest() {
    whenever(request.requestURL).thenReturn(StringBuffer("http://some.com:443/where"))
    whenever(request.queryString).thenReturn("param=value")
    whenever(request.isSecure).thenReturn(true)
    cache.saveRequest(request, response)
    verify(helper).addCookieToResponse(
      same(request),
      same(response),
      check {
        val url = String(Base64Utils.decodeFromString(it))
        assertThat(url).isEqualTo("https://some.com/where?param=value")
      }
    )
  }

  @Test
  fun saveRequest_insecureRequest() {
    whenever(request.requestURL).thenReturn(StringBuffer("https://some.com/where"))
    whenever(request.queryString).thenReturn("param=value")
    whenever(request.isSecure).thenReturn(false)
    cache.saveRequest(request, response)
    verify(helper).addCookieToResponse(
      same(request),
      same(response),
      check {
        val url = String(Base64Utils.decodeFromString(it))
        assertThat(url).isEqualTo("http://some.com/where?param=value")
      }
    )
  }

  @Test
  fun saveRequest_differentPort() {
    whenever(request.requestURL).thenReturn(StringBuffer("https://some.com:12345/where"))
    whenever(request.queryString).thenReturn("param=value")
    whenever(request.isSecure).thenReturn(true)
    cache.saveRequest(request, response)
    verify(helper).addCookieToResponse(
      same(request),
      same(response),
      check {
        val url = String(Base64Utils.decodeFromString(it))
        assertThat(url).isEqualTo("https://some.com:12345/where?param=value")
      }
    )
  }

  @Test
  fun getRequest() {
    whenever(helper.readValueFromCookie(request)).thenReturn(
      Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".toByteArray()))
    )
    val savedRequest = cache.getRequest(request, response)
    assertThat(savedRequest?.redirectUrl).isEqualTo("https://some.com/where?param=value")
  }

  @Test
  fun matchingRequest_matches() {
    whenever(request.requestURL).thenReturn(StringBuffer("http://some.com:443/where"))
    whenever(request.queryString).thenReturn("param=value")
    whenever(request.isSecure).thenReturn(true)
    whenever(helper.readValueFromCookie(request)).thenReturn(
      Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".toByteArray()))
    )
    val savedRequest = cache.getMatchingRequest(request, response)
    assertThat(savedRequest).isSameAs(request)
    verify(helper).removeCookie(request, response)
  }

  @Test
  fun matchingRequest_nomatch() {
    whenever(request.requestURL).thenReturn(StringBuffer("http://some.com:443/where"))
    whenever(request.queryString).thenReturn("param=othervalue")
    whenever(request.isSecure).thenReturn(true)
    whenever(helper.readValueFromCookie(request)).thenReturn(
      Optional.of(Base64Utils.encodeToString("https://some.com/where?param=value".toByteArray()))
    )
    val savedRequest = cache.getMatchingRequest(request, response)
    assertThat(savedRequest).isNull()
    verify(helper, never()).removeCookie(request, response)
  }

  @Test
  fun matchingRequest_noSavedRequest() {
    whenever(helper.readValueFromCookie(request)).thenReturn(Optional.empty())
    val savedRequest = cache.getMatchingRequest(request, response)
    assertThat(savedRequest).isNull()
  }
}
